lib/redirects.js

```diff
@@ -2,58 +2,90 @@
 
 const gw = require('./gateways');
 
-var {
-  Cc, Ci
-} = require('chrome');
+var { Cc, Ci, Cr } = require('chrome');
+
+const protocols = require("./protocols.js")
 
 var ioservice = Cc['@mozilla.org/network/io-service;1'].getService(Ci.nsIIOService);
+var events = require("sdk/system/events");
 
-var ipfsRequestObserver = {
-  observe: function(subject, topic, data) { // jshint unused:false
-    if (topic == 'http-on-modify-request') {
-      let channel = subject.QueryInterface(Ci.nsIHttpChannel);
-      let httpUrl = channel.URI.spec;
-      if (httpUrl.match(gw.publicHosts()) && httpUrl.match(gw.IPFS_RESOURCE)) {
-        channel.setRequestHeader('x-ipfs-firefox-addon', 'true', false);
-        if (gw.isEnabled()) {
-          //console.info('Detected HTTP request to the public gateway: ' + channel.URI.spec);
-          let uri = ioservice.newURI(httpUrl.replace(gw.publicHosts(), gw.customUri().spec), null, null);
-          //console.info('Redirecting to custom gateway: ' + uri.spec);
-          channel.redirectTo(uri);
-        }
-      }
-    }
-  },
-  get observerService() {
-    return Cc['@mozilla.org/observer-service;1'].getService(Ci.nsIObserverService);
-  },
-  register: function() {
-    if (this.registered)
-      return;
-    this.registered = true;
-    this.observerService.addObserver(this, 'http-on-modify-request', false);
-  },
+const listener = function(event) {
+  //console.log("obs", event);
+  let {subject, type, data} = event;
 
-  unregister: function() {
-    if (!this.registered)
+  if (type == 'http-on-modify-request') {
+    if(!gw.isEnabled() && prefs.rewriteAs == "gateway")
       return;
-    this.registered = false;
-    this.observerService.removeObserver(this, 'http-on-modify-request');
+    let channel = subject.QueryInterface(Ci.nsIHttpChannel);
+    let httpUrl = channel.URI.spec;
+    let isGatewayRequest = !!(httpUrl.match(gw.publicHosts()) && httpUrl.match(gw.IPFS_RESOURCE));
+    // TODO: handle gateway -> local redirects
+    if(!isGatewayRequest)
+      return;
+    let ipfsURI = ioservice.newURI(protocols.fsScheme + ":" + channel.URI.path, null, null);
+    channel.setRequestHeader('x-ipfs-firefox-addon', 'true', false);
+
+    //console.log("comp", ipfsURI.spec, channel.originalURI.spec, ipfsURI.spec == channel.originalURI.spec)
+
+    if(prefs.useCustomGateway || !channel.originalURI || channel.originalURI.spec != ipfsURI.spec ) {
```

Comment:
`redirects.js:30`: we probably want to use `gw.isEnabled()` here as  `ReferenceError: prefs is not defined` (I got this error while testing b5d785b)